### DISCUSSION TITLE
Progressive blur with transparency

### DISCUSSION THREAD

**Discussion by TheEndless** [ImageMagick/ImageMagick#2420] [Help] [answered]
I'm trying to apply a progressive blur to an image with transparency, but the result replaces the transparency with black.  I'm sure it's something silly that I'm missing, but what is it?  A simple example...

```
convert -size 600x600 xc:white -background None \
    -gravity center -extent 700x700 \
    \( -size 700x700 gradient:gray5-white \) \
    -compose Blur -set option:compose:args 10 \
    -composite blurred.png
```

Thanks in advance!

**Comment by fmw42** [comment]
On Unix-like systems, parentheses must be escaped. Try

```
convert -size 600x600 xc:white -background None \
-gravity center -extent 700x700 \
\( -size 700x700 gradient:gray5-white \) \
-compose Blur -set option:compose:args 10 \
-composite blurred.png

```

**Reply by TheEndless** [reply]
sorry.. I tried it as one line.. broken escaping is in my attempt to translate it here only :P  (corrected.. same result)

**Reply by fmw42** [reply]
If one line, remove the ending \s.  What is your IM version and platform? If IM 7, use magick rather than convert. If Windows, the syntax is different.

**Reply by TheEndless** [reply]
apologies.. I'm using Magick.net.. decided to switch to IM directly to see if I got the same results before asking here.  I do get the same results... using IM7.

For clarity, this is the full command-line on Windows:
```magick -size 600x600 xc:white -background None -gravity center -extent 700x700 ( -size 700x700 gradient:gray5-white ) -compose Blur -set option:compose:args 10 -composite blurred.png```

results in the following:
<img src="https://i.imgur.com/9GNjTIU.png" />

Happy to translate the native IM command-line to Magick.net myself.  Just need to know what I'm missing to make that `-compose Blur` retain the transparency, without it blending to black.

**Reply by fmw42** [reply]
In IM 7, use magick, not magick convert. All the rest of the tools need magick, such as magick identify.

**Reply by TheEndless** [reply]
Same result.

**Reply by fmw42** [reply]
Generally, it won't matter. But it is then using ImageMagick 6. If you want to use IM 7 features, then use only magick.

**Reply by TheEndless** [reply]
Understood, but that's not where the issue is.  Again, I only put together an IM command line to illustrate what I'm trying to accomplish. The intent was to provide a simple example, but apparently I just confused things.

What I'm trying to accomplish is applying a progressive blur without losing the transparency.  Is that possible?  In the example above, I used a solid white square, for simplicity, but the actual effect will be applied to a full color image.

**Reply by fmw42** [reply]
I suspect you will need to save the alpha channel, do your processing on the RGB channels, then re-apply the alpha channel.

**Reply by TheEndless** [reply]
I tried that, actually.  It's closer, but results in the edge being gray as those pixels were blended with black during the blur operation.

![](https://i.imgur.com/Qz44Ip9.png)

If I retain the alpha channel without blurring it as well, it results in the same gray, but with a hard edge...
 ![](https://i.imgur.com/m8OvBOa.png)

**Comment by snibgo** [comment]
What version of IM are you using?

Using v6.9.9-50, the result has transparency. All pixels that have any opacity are white. The colour behind the transparency is ignored for the purpose of the blur, so the only effect is to blur the alpha. I suppose this is the result you want.

Using v7.0.10-10, the result have an alpha channel but that is 100% (fully opaque) for all pixels.

I suspect a bug in v7.0.10-10.

For this image, you could make a black and white image, blur that, then CopyOpacity that to an entirely white image.

**Reply by TheEndless** [reply]
I'm using Magick.Net 7.21.1, which I assume is based on IM v7.something.  The sample images above were generated with IM 7.0.10-27 just to illustrate the issue without limiting it to my experience with Magick.Net.

Unfortunately, the actual images I'm working with are full color.  I used white here for simplicity and to make the effect more obvious.  I'm basically doing the same thing, though.. generating a black and white mask with the same blur and applying it to the final image.  The end result isn't terrible, but the gray pixels are much more noticeable in some images than others.

I'll give an older 6.x version of Magick.Net a try and see if it produces different results.

**Reply by TheEndless** [reply]
That did the trick!  Switching to Magick.Net v6.8.9.601 (the last 6.x release) gives me the expected results.  All 7.x versions I tried gave me the undesired results, so definitely looks like a 7.x change... whether it's a bug or expected, I couldn't say, but at least I feel better about not being able to figure out what I was doing wrong 😛 ..  Thanks for the tip, @snibgo!

**Comment by snibgo** [comment]
V7 has changed some processing of alpha. But I don't see why it should make all pixels opaque in this example.

A developer may comment.

**Comment by TheEndless** [comment]
Just going to bump this (is that a thing on github?) in hopes that a dev might add insight...  unfortunately, rolling back to the older version of Magick.Net turned out to be a non-starter, as I'm using .Net core and couldn't get it working with my website.

**Reply by dlemstra** [reply]
I will try to do that this weekend. I did not realize you were waiting for this otherwise I would have released a new version earlier.

**Reply by TheEndless** [reply]
excellent.  thank you!

**Comment by snibgo** [comment, marked as answer]
Yeah, bumping is a thing.

I have narrowed the problem down to "-compose blur -composite", which has changed in v7 compared to v6. In v7, pixels that were transparent become opaque. Worse, the colours of pixels that were fully transparent are blurred into other pixels, so there is no workaround. 

The problem does not occur for a simple "-blur 0x5" operation.

In my view, this is a bug.

**Reply by dlemstra** [reply]
I just pushed a patch to fix this but I need to make sure that this is the correct fix. Wanted to wait with an answer till I had a confirmation that my patch was correct.

**Reply by TheEndless** [reply]
@dlemstra any idea when this patch will make it into the Magick.Net Nuget package?  Thanks for looking into it!

**Reply by dlemstra** [reply]
The new release has just been published.

**Comment by GeeMack** [comment]
I have tried the command from the OP, adapted for Windows, with both "IM 6.9.11-29 Q16 x64" and "IM 7.0.10-29 Q16 x64". The results from both have the expected graduated blur.

    convert -size 600x600 xc:darkred -background None ^
        -gravity center -extent 700x700 ^
        ( -size 700x700 gradient:gray5-white ) ^
        -compose Blur -set option:compose:args 10 ^
        -composite blurred_v6.png

![blurred_v6](https://user-images.githubusercontent.com/65475279/92966098-87190380-f43c-11ea-90b9-db4229d7d4a0.png)

**Reply by TheEndless** [reply]
yeah, it works in pre-7.0 builds and was patched in 7.0.10-29 (or 28).  thanks for confirming that it works as expected now!
